core/domain/feature_flag_services.py

```diff
@@ -0,0 +1,282 @@
+    """Exception thrown when an unknown feature flag is requested."""
+
+    pass
+
+
+def update_feature_flag(
+    feature_flag_name: str,
+    force_enable_for_all_users: bool,
+    rollout_percentage: int,
+    user_group_ids: List[str],
+) -> None:
+    """Updates the feature flag.
+
+    Args:
+        feature_flag_name: str. The name of the feature flag to update.
+        force_enable_for_all_users: bool. Whether the feature flag is
+            force-enabled for all the users.
+        rollout_percentage: int. The percentage of logged-in users for which
+            the feature will be enabled. This value is ignored if the
+            force_enable_for_all_users property is set to True.
+        user_group_ids: List[str]. The list of ids of UserGroup objects.
+
+    Raises:
+        FeatureFlagNotFoundException. Feature flag trying to update does
+            not exist.
+    """
+    if feature_flag_name not in ALL_FEATURES_NAMES_SET:
+        raise FeatureFlagNotFoundException(
+            'Unknown feature flag: %s.' % feature_flag_name)
+
+    registry.Registry.update_feature_flag(
+        feature_flag_name,
+        force_enable_for_all_users,
+        rollout_percentage,
+        user_group_ids
+    )
+
+
+def get_all_feature_flags() -> List[feature_flag_domain.FeatureFlag]:
+    """Returns all feature flags. This method is used for providing detailed
+    feature flags information to the release coordinator page.
+
+    Returns:
+        feature_flags: list(FeatureFlag). A list containing the dict mappings
+        of all fields of the feature flags.
+
+    Raises:
+        Exception. Feature flag does not exists.
+    """
+    feature_flags: List[feature_flag_domain.FeatureFlag] = []
+    feature_flags_to_fetch_from_storage = []
+
+    for feature_flag_name_enum in ALL_FEATURE_FLAGS:
+        if feature_flag_name_enum.value not in (
+            registry.Registry.feature_flag_spec_registry
+        ):
+            raise Exception(
+                'Feature flag not found: %s.' % feature_flag_name_enum.value)
+
+        feature_flags_to_fetch_from_storage.append(
+            feature_flag_name_enum.value)
+
+    feature_flags_from_storage = load_feature_flags_from_storage(
+        feature_flags_to_fetch_from_storage)
+
+    for feature_flag_name, feature_flag in (
+        feature_flags_from_storage.items()
+    ):
+        if feature_flag is not None:
+            feature_flag_spec = (
+                registry.Registry.feature_flag_spec_registry[
+                    feature_flag.name])
+            # Rulling out the possibility of last_updated to be None as we
+            # are getting the feature flag value from the storage.
+            assert feature_flag.feature_flag_config.last_updated is not None
+            last_updated = utils.convert_naive_datetime_to_string(
+                feature_flag.feature_flag_config.last_updated)
+            feature_flag = feature_flag_domain.FeatureFlag.from_dict({
```

Comment:
This is repeated below. Have a bit less duplication -- move this and the next statement out of the if/else, and define variables enable_for_all_users, rollout_percentage, etc. instead in both branches. Then you only need to create the dict once.

Also, same comment as elsewhere, don't use from_dict() unless necessary, prefer the construtor of the plain python object.

Reply:
Updated